backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/query/impl/LuceneSearchResultExtractorImpl.java

```diff
@@ -88,23 +101,57 @@
 		return extractedData;
 	}
 
-	private void addChildrenContributions(IndexSearcher indexSearcher) throws IOException {
-		Document document = storedFieldVisitor.getDocument();
-		String id = document.getField( LuceneFields.idFieldName() ).stringValue();
-		ScoreDoc[] children = findChildren( indexSearcher, id );
-		for ( ScoreDoc childHit : children ) {
-			indexSearcher.doc( childHit.doc , storedFieldVisitor );
+	private Map<Integer, Set<Integer>> fetchNestedDocs(IndexSearcher indexSearcher, ScoreDoc[] scoreDocs) throws IOException {
+		// if the projection does not need any nested object skip their fetching
+		if ( !rootProjection.involveNestedFields() ) {
+			return new HashMap<>();
+		}
+
+		Map<String, Integer> parentIds = new HashMap<>();
+		for ( ScoreDoc hit : scoreDocs ) {
+			Document doc = indexSearcher.doc( hit.doc, ID_FIELD_SET );
+			String parentId = doc.getField( LuceneFields.idFieldName() ).stringValue();
+			if ( parentId == null ) {
+				continue;
+			}
+			parentIds.put( parentId, hit.doc );
+		}
+
+		Map<String, Set<Integer>> stringSetMap = fetchChildren( indexSearcher, parentIds.keySet() );
+		HashMap<Integer, Set<Integer>> result = new HashMap<>();
+		for ( Map.Entry<String, Set<Integer>> entry : stringSetMap.entrySet() ) {
+			result.put( parentIds.get( entry.getKey() ), entry.getValue() );
 		}
+		return result;
 	}
 
-	public ScoreDoc[] findChildren(IndexSearcher indexSearcher, String id) {
-		TermQuery query = new TermQuery( new Term( LuceneFields.rootIdFieldName(), id ) );
+	private Map<String, Set<Integer>> fetchChildren(IndexSearcher indexSearcher, Set<String> parentIds) {
+		BooleanQuery.Builder builder = new BooleanQuery.Builder();
+		for ( String parentId : parentIds ) {
+			TermQuery query = new TermQuery( new Term( LuceneFields.rootIdFieldName(), parentId ) );
+			builder.add( query, BooleanClause.Occur.SHOULD );
+		}
+		builder.setMinimumNumberShouldMatch( 1 );
+
+		BooleanQuery booleanQuery = builder.build();
+		Map<String, Set<Integer>> result = new HashMap<>();
 
```

Comment:
On a related note, it's a shame that we are collecting *all* children, whereas we really only need *some* of them. Namely, we need the nested documents that hold the data of the fields we're projecting on.

There could be many "nested object fields" in a single document, with many different paths: "authors", "categories", "authors.publishers", ... If we only project on "authors.name", why should we also fetch "categories" and "authors.publishers"?

By switching to `ToChildBlockJoinQuery` as mentioned above, we could use a query to filter child documents by their path : something like `new TermQuery( new Term( LuceneFields.nestedDocumentPathFieldName(), "<the path of the nested document we're interested in" )`. Which could potentially reduce the number of fetched children in many cases.

Of course, that would require to know the paths we're interested in. But that shouldn't be too hard: currently we store the information "do we need nested documents", as a boolean; we only have to turn this into a set of paths we're interested in. These paths could be passed through `org.hibernate.search.backend.lucene.search.projection.impl.LuceneSearchProjection#contributeFields`, I suspect.